crates/futures/src/lib.rs

```diff
@@ -353,8 +353,32 @@ fn _future_to_promise(future: Box<Future<Item = JsValue, Error = JsValue>>) -> P
                 // continue polling. For us, though, there's nothing else to do,
                 // so we bail out.
                 // later see
-                State::Polling => {}
-            }
+                State::Polling => return,
+            };
+
+            // Use `setTimeout` to place our execution onto the next turn of the
+            // event loop, enqueueing our poll operation. We don't currently
```

Comment:
I don't think we want to use `setTimeout` -- that will incure larger-than-necessary latency.

Instead, we should do the equivalent of `Promise.resolve().then(() => executeOurStuff())` which will execute on the micro-task queue, and not bounce all the way out to the browser's main event loop.

Reply:
In stdweb [we use `MutationObserver` (falling back to `Promise.prototype.then` on Nodejs)](https://github.com/koute/stdweb/blob/bc07dac9d275e6d0b517d304527d3a593df208f6/src/webcore/executor.rs#L268-L310).

The reason we do this is because `Promise.prototype.then` allocates (to create the new Promise), whereas with `MutationObserver` we can schedule the next microtask tick without any allocation (well, except the allocation for the array of `MutationRecord`, I guess).

We also handle everything in Rust as much as possible:

1. It [maintains a `VecDeque` of pending tasks](https://github.com/koute/stdweb/blob/bc07dac9d275e6d0b517d304527d3a593df208f6/src/webcore/executor.rs#L111).

2. When there are multiple pending tasks, [it ensures that it only schedules a single microtask tick](https://github.com/koute/stdweb/blob/bc07dac9d275e6d0b517d304527d3a593df208f6/src/webcore/executor.rs#L220-L228).

3. When the microtask tick finishes, [it then loops over the pending tasks](https://github.com/koute/stdweb/blob/bc07dac9d275e6d0b517d304527d3a593df208f6/src/webcore/executor.rs#L165).

This ensures that no matter how many tasks are pending, it only needs to schedule one microtask tick, which made a *huge* difference for performance (the old implementation was similar to wasm-bindgen and spun up a new microtask tick for each task).

There's a few reasons for the big performance boost:

* By keeping everything in Rust as much as possible, it can make use of Rust's great data structures and optimizations.

* Calling between wasm and JS is slow (at least in stdweb), so this minimizes the amount of wasm<->JS marshalling (because it only schedules one microtask tick per cycle, rather than one microtask tick per task).

* It cuts back massively on the amount of garbage that JS produced (which was causing a lot of garbage collector thrashing).

   We're talking huge differences here: with [my animation demo](https://pauan.github.io/rust-dominator/examples/animation/) the old executor got ~10 FPS (and generated a lot of garbage), but [when I changed it to use a `VecDeque`](https://github.com/koute/stdweb/pull/165) it jumped up to ~60 FPS (and generated almost no garbage).

In any case, I agree that it should be using the microtask queue (`MutationRecord` or Promises).

Reply:
Oh good idea, thanks for the info @fitzgen! @Pauan I think we can eventually go towards a more complicated world like that, but I'd prefer to have a benchmark and/or example which motivates it to avoid too much premature complexity.